lib/ansible/modules/windows/win_blockinfile.ps1

```diff
@@ -0,0 +1,330 @@
+                If ($preamble[$i] -ne $bom[$i]) {
+                    break;
+                }
+                ElseIf ($i + 1 -eq $preamble.Length) {
+                    $encodingObject = $encoding;
+                    $found = $true;
+                }
+            }
+            If ($found) {
+                break;
+            }
+        }
+    }
+    return $encodingObject;
+}
+
+function BackupFile {
+    param(
+        [String]$path,
+        [bool]$check_mode
+    );
+
+    $backuppath = $path + "." + [DateTime]::Now.ToString("yyyyMMdd-HHmmss");
+    Try {
+        Copy-Item $path $backuppath -WhatIf:$check_mode;
+    }
+    Catch {
+        $module.FailJson("Cannot copy backup file! ($($_.Exception.Message))", $_);
+    }
+    If (-not $check_mode) {
+        Try {
+            Get-Acl -Path $path | Set-Acl -Path $backuppath
+        }
+        Catch {
+            $module.FailJson("Cannot copy ACL to backup file! ($($_.Exception.Message))", $_);
+        }
+    }
+    return $backuppath;
+}
+
+If (Test-Path -LiteralPath $path -PathType "container") {
+    $module.FailJson("Path $path is a directory", $_);
+}
+
+$linesep = "`r`n"
+If ($newline -eq "unix") {
+    $linesep = "`n";
+}
+
+If ($insertbefore -and $insertafter) {
+    $module.Warn("Both insertbefore and insertafter parameters found, ignoring `"insertafter=$insertafter`"")
+}
+
+If (-not $insertbefore -and -not $insertafter) {
+    $insertafter = "EOF";
+}
+
+$insre = $null;
+If ($insertafter -and $insertafter -ne "BOF" -and $insertafter -ne "EOF") {
+    $insre = New-Object Regex $insertafter, 'Compiled';
+}
+ElseIf ($insertbefore -and $insertbefore -ne "BOF" -and $insertbefore -ne "EOF") {
+    $insre = New-Object Regex $insertbefore, 'Compiled';
+}
+
+
+# The default encoding is UTF-8 without BOM
+$encodingobj = [System.Text.UTF8Encoding] $false;
+
+If ($encoding -ne "auto") {
+    $encodingobj = [System.Text.Encoding]::GetEncoding($encoding);
+}
+ElseIf (Test-Path -LiteralPath $path) {
+    $encodingobj = GuessFileEncoding $path
+
+}
+
+$module.Result.backup = ""
+$module.Result.changed = $false
+$module.Result.msg = ""
+
+$cleanpath = $path.Replace("/", "\");
+If (-not (Test-Path -LiteralPath $path)) {
+    If ($state -eq "absent") {
+        $module.Result.changed = $false
+        $module.Result.msg = "File does not exist"
+        $module.ExitJson()
+    }
+    If (-not $create) {
+        $module.FailJson("Path $path does not exist and create not specified", $_);
+    }
+    $lines = New-Object System.Collections.ArrayList($null)
+}
+Else {
+    # Read the dest file lines using the indicated encoding into a mutable ArrayList. Note
+    # that we have to clean up the path because ansible wants to treat / and \ as
+    # interchangeable in windows pathnames, but .NET framework internals do not support that.
+
+    # We read the file using ReadAllText in order to preserve ending newlines
+    $origContent = [System.IO.File]::ReadAllText($cleanpath, $encodingobj);
```

Comment:
Why not [`Get-Content -Path $cleanpath -Encoding $encodeing`](https://docs.microsoft.com/en-us/powershell/module/Microsoft.PowerShell.Management/Get-Content?view=powershell-3.0)?

Reply:
See [here](https://github.com/ansible/ansible/pull/52586#discussion_r258712639).